File: flask-app/app/services.py
from datetime import date, timedelta

from app import db
from app.models import ChecklistDiario, Rotina, RegistroEstudo, DIAS_SEMANA, DIAS_SEMANA_ABREV


def meta_minutos_dia(dia_semana):
    """Soma a duração de todos os blocos de tipo 'estudo' de um dia da semana (0=Segunda)."""
    total = 0
    for r in Rotina.query.filter_by(dia_semana=dia_semana, tipo="estudo").all():
        total += r.duracao
    return total


def meta_semanal_minutos():
    return sum(meta_minutos_dia(d) for d in range(7))


def garantir_checklist_do_dia(usuario_id, data_ref):
    """Cria (se ainda não existirem) os itens de checklist do dia a partir da Rotina."""
    dia_semana = data_ref.weekday()  # 0=Segunda ... 6=Domingo
    existentes = ChecklistDiario.query.filter_by(usuario_id=usuario_id, data=data_ref).count()
    if existentes > 0:
        return

    rotinas = Rotina.query.filter_by(dia_semana=dia_semana).order_by(Rotina.ordem).all()
    for r in rotinas:
        item = ChecklistDiario(
            usuario_id=usuario_id,
            data=data_ref,
            atividade=r.atividade,
            tipo=r.tipo,
            ordem=r.ordem,
            concluido=False,
        )
        db.session.add(item)
    db.session.commit()


def progresso_dia(usuario_id, data_ref):
    itens = ChecklistDiario.query.filter_by(usuario_id=usuario_id, data=data_ref).all()
    total = len(itens)
    concluidos = sum(1 for i in itens if i.concluido)
    pct = int(round((concluidos / total) * 100)) if total else 0
    return concluidos, total, pct


def minutos_estudados_no_dia(usuario_id, data_ref):
    registros = RegistroEstudo.query.filter_by(usuario_id=usuario_id, data=data_ref).all()
    return sum(r.minutos for r in registros)


def inicio_da_semana(data_ref):
    """Retorna a segunda-feira da semana de data_ref."""
    return data_ref - timedelta(days=data_ref.weekday())


def resumo_semana(usuario_id, data_ref):
    seg = inicio_da_semana(data_ref)
    dias = []
    total_estudado = 0
    total_meta = 0
    for i in range(7):
        d = seg + timedelta(days=i)
        meta = meta_minutos_dia(i)
        estudado = minutos_estudados_no_dia(usuario_id, d)
        concluido = estudado >= meta and meta > 0
        dias.append({
            "data": d,
            "nome": DIAS_SEMANA[i],
            "abrev": DIAS_SEMANA_ABREV[i],
            "meta_minutos": meta,
            "estudado_minutos": estudado,
            "concluido": concluido,
            "eh_hoje": d == date.today(),
        })
        total_estudado += estudado
        total_meta += meta
    pct = int(round((total_estudado / total_meta) * 100)) if total_meta else 0
    return {
        "dias": dias,
        "total_estudado": total_estudado,
        "total_meta": total_meta,
        "pct": min(pct, 100),
        "inicio": seg,
        "fim": seg + timedelta(days=6),
    }


def sequencia_dias(usuario_id):
    """Conta quantos dias seguidos (a partir de hoje, para trás) a meta de estudo foi batida."""
    dias = 0
    d = date.today()
    while True:
        meta = meta_minutos_dia(d.weekday())
        estudado = minutos_estudados_no_dia(usuario_id, d)
        if meta > 0 and estudado >= meta:
            dias += 1
            d = d - timedelta(days=1)
        else:
            break
    return dias


def total_questoes(usuario_id):
    registros = RegistroEstudo.query.filter_by(usuario_id=usuario_id, tipo="Questões").all()
    return len(registros)


def fmt_minutos(minutos):
    h = minutos // 60
    m = minutos % 60
    if h and m:
        return f"{h}h{m:02d}"
    if h:
        return f"{h}h"
    return f"{m}min"


NOME_DIA_PT = ["Segunda-feira", "Terça-feira", "Quarta-feira", "Quinta-feira", "Sexta-feira", "Sábado", "Domingo"]

File: src/services.ts
import type { D1Database } from "@cloudflare/workers-types";
import * as db from "./db";

export const DIAS_SEMANA = ["Segunda", "Terça", "Quarta", "Quinta", "Sexta", "Sábado", "Domingo"];
export const NOME_DIA_PT = [
  "Segunda-feira",
  "Terça-feira",
  "Quarta-feira",
  "Quinta-feira",
  "Sexta-feira",
  "Sábado",
  "Domingo",
];
export const TIPOS_ESTUDO = ["Teoria", "Questões", "Revisão", "Anki", "Redação"];

// data no formato ISO "YYYY-MM-DD", sempre em relação ao fuso horário de São Paulo.
export function hojeISO(): string {
  const fmt = new Intl.DateTimeFormat("en-CA", { timeZone: "America/Sao_Paulo" });
  return fmt.format(new Date()); // en-CA -> YYYY-MM-DD
}

export function diaSemanaDe(dataISO: string): number {
  // 0=Segunda ... 6=Domingo
  const [y, m, d] = dataISO.split("-").map(Number);
  const jsDay = new Date(Date.UTC(y, m - 1, d)).getUTCDay(); // 0=Domingo ... 6=Sábado
  return (jsDay + 6) % 7;
}

export function somaDias(dataISO: string, dias: number): string {
  const [y, m, d] = dataISO.split("-").map(Number);
  const dt = new Date(Date.UTC(y, m - 1, d));
  dt.setUTCDate(dt.getUTCDate() + dias);
  return dt.toISOString().slice(0, 10);
}

export function inicioDaSemana(dataISO: string): string {
  const diaSemana = diaSemanaDe(dataISO);
  return somaDias(dataISO, -diaSemana);
}

export function formatarDataBR(dataISO: string): string {
  const [y, m, d] = dataISO.split("-");
  return `${d}/${m}/${y}`;
}

export function formatarDataBRCurta(dataISO: string): string {
  const [, m, d] = dataISO.split("-");
  return `${d}/${m}`;
}

export async function metaMinutosDia(database: D1Database, usuarioId: number, diaSemana: number): Promise<number> {
  const rotinas = await db.listRotinaPorDiaETipo(database, usuarioId, diaSemana, "estudo");
  return rotinas.reduce((acc, r) => acc + db.duracaoRotina(r), 0);
}

export async function garantirChecklistDoDia(database: D1Database, usuarioId: number, dataRef: string): Promise<void> {
  const existentes = await db.contarChecklistDoDia(database, usuarioId, dataRef);
  if (existentes > 0) return;

  const diaSemana = diaSemanaDe(dataRef);
  const rotinas = await db.listRotinaPorDia(database, usuarioId, diaSemana);
  for (const r of rotinas) {
    await db.criarChecklistItem(database, usuarioId, dataRef, r.atividade, r.tipo, r.ordem);
  }
}

export async function progressoDia(
  database: D1Database,
  usuarioId: number,
  dataRef: string
): Promise<{ concluidos: number; total: number; pct: number }> {
  const itens = await db.listChecklistDoDia(database, usuarioId, dataRef);
  const total = itens.length;
  const concluidos = itens.filter((i) => i.concluido).length;
  const pct = total ? Math.round((concluidos / total) * 100) : 0;
  return { concluidos, total, pct };
}

export interface DiaResumo {
  data: string;
  nome: string;
  metaMinutos: number;
  estudadoMinutos: number;
  concluido: boolean;
  ehHoje: boolean;
}

export interface ResumoSemana {
  dias: DiaResumo[];
  totalEstudado: number;
  totalMeta: number;
  pct: number;
  inicio: string;
  fim: string;
}

export async function resumoSemana(database: D1Database, usuarioId: number, dataRef: string): Promise<ResumoSemana> {
  const seg = inicioDaSemana(dataRef);
  const hoje = hojeISO();
  const dias: DiaResumo[] = [];
  let totalEstudado = 0;
  let totalMeta = 0;

  for (let i = 0; i < 7; i++) {
    const d = somaDias(seg, i);
    const meta = await metaMinutosDia(database, usuarioId, i);
    const estudado = await db.minutosEstudadosNoDia(database, usuarioId, d);
    const concluido = meta > 0 && estudado >= meta;
    dias.push({ data: d, nome: DIAS_SEMANA[i], metaMinutos: meta, estudadoMinutos: estudado, concluido, ehHoje: d === hoje });
    totalEstudado += estudado;
    totalMeta += meta;
  }

  const pct = totalMeta ? Math.round((totalEstudado / totalMeta) * 100) : 0;
  return {
    dias,
    totalEstudado,
    totalMeta,
    pct: Math.min(pct, 100),
    inicio: seg,
    fim: somaDias(seg, 6),
  };
}

export async function sequenciaDias(database: D1Database, usuarioId: number): Promise<number> {
  let dias = 0;
  let d = hojeISO();
  while (true) {
    const meta = await metaMinutosDia(database, usuarioId, diaSemanaDe(d));
    const estudado = await db.minutosEstudadosNoDia(database, usuarioId, d);
    if (meta > 0 && estudado >= meta) {
      dias += 1;
      d = somaDias(d, -1);
    } else {
      break;
    }
  }
  return dias;
}

export function fmtMinutos(minutos: number): string {
  const h = Math.floor(minutos / 60);
  const m = minutos % 60;
  if (h && m) return `${h}h${String(m).padStart(2, "0")}`;
  if (h) return `${h}h`;
  return `${m}min`;
}

export function parseTempo(texto: string): number {
  const t = (texto || "").trim().toLowerCase().replace(/\s+/g, "");
  if (!t) return 0;
  if (t.includes("h")) {
    const [horasStr, minutosStr] = t.split("h");
    const horas = horasStr ? parseInt(horasStr, 10) : 0;
    const minutos = minutosStr ? parseInt(minutosStr, 10) : 0;
    return (Number.isNaN(horas) ? 0 : horas) * 60 + (Number.isNaN(minutos) ? 0 : minutos);
  }
  if (t.includes(":")) {
    const [h, m] = t.split(":");
    return (parseInt(h, 10) || 0) * 60 + (parseInt(m, 10) || 0);
  }
  const n = parseInt(t, 10);
  return Number.isNaN(n) ? 0 : n;
}
